src/integrationtests/java/com/aws/greengrass/integrationtests/logmanager/LogManagerTest.java

```diff
@@ -453,4 +466,162 @@ void GIVEN_system_config_with_small_periodic_interval_WHEN_active_logs_included_
+        // size should be contents of 4 files.
+        verify(cloudWatchLogsClient, times(1)).putLogEvents(captor.capture());
+
+        List<PutLogEventsRequest> putLogEventsRequests = captor.getAllValues();
+        assertEquals(1, putLogEventsRequests.size());
+        for (PutLogEventsRequest request : putLogEventsRequests) {
+            assertEquals(calculateLogStreamName(THING_NAME, LOCAL_DEPLOYMENT_GROUP_NAME), request.logStreamName());
+            assertEquals("/aws/greengrass/UserComponent/" + AWS_REGION + "/UserComponentB", request.logGroupName());
+            assertNotNull(request.logEvents());
+            assertEquals(DEFAULT_LOG_LINE_IN_FILE * testFileNumber, request.logEvents().size());
+            assertEquals(DEFAULT_FILE_SIZE * testFileNumber,
+                    request.logEvents().stream().mapToLong(value -> value.message().length()).sum());
+        }
+        // confirm no file is deleted
+        Pattern logFileNamePattern = Pattern.compile("\\w*UserComponentB\\w*.\\w*");
+        LogFileGroup logFileGroup = LogFileGroup.create(logFileNamePattern, tempDirectoryPath.toUri(), mockInstant);
+        assertEquals(testFileNumber, logFileGroup.getLogFiles().size());
+    }
+
+    @Test
+    void GIVEN_files_randomly_named_WHEN_all_files_renamed_randomly_and_new_data_in_active_file_THEN_all_files_only_send_once()
+            throws Exception {
+        //TODO: Intentionally pass the test now, since it will fail for current code.
+        if (!ACTIVE_LOG_FILE_FEATURE_ENABLED_FLAG.get()) {
+            return;
+        }
+        when(cloudWatchLogsClient.putLogEvents(any(PutLogEventsRequest.class))).thenReturn(
+                PutLogEventsResponse.builder().nextSequenceToken("nextToken").build());
+
+        tempDirectoryPath = Files.createDirectory(tempRootDir);
+        Instant mockInstant = Instant.EPOCH;
+        LogConfig.getRootLogConfig().setLevel(Level.TRACE);
+        LogConfig.getRootLogConfig().setStore(LogStore.FILE);
+        LogManager.getLogConfigurations().putIfAbsent("UserComponentB",
+                newLogConfigFromRootConfig(LogConfigUpdate.builder().fileName("UserComponentB.log").build()));
+
+        List<LogFile> logFiles = new ArrayList<>();
+        int testFileNumber = 4;
+        for (int i = 0; i < testFileNumber; i++) {
+            String randomName1 = UUID.randomUUID().toString();
+            LogFile logFile = createFileAndWriteData(tempDirectoryPath,
+                    randomName1 + "UserComponentB.");
+            logFiles.add(logFile);
+        }
+
+        setupKernel(tempDirectoryPath, "randomlyNamedFilesWithoutComponentsConfig.yaml");
+        assertTrue(isActiveFileProcessedOrTimeout(logManagerService));
+
+        // randomly rename file and see if uploader trying to process more file.
+        logFiles.sort(Comparator.comparingLong(LogFile::lastModified));
+        for (int i = 0; i < logFiles.size(); i++) {
```

Comment:
This could be a method that renames the files and you could make it more declarative and refactor it. It is hard to figure out what the intent of this is by reading it.

Also the loop is doing 2 things renaming and writing to the active file. This could be split into

```java
rotateLogFiles(logFiles);
LogFile activeLogFile = logFiles.get(logFiles.size() - 1);
writeLogsToFile(activeLogFile)
```